lib/puppet/provider/onesecgroup/cli.rb

```diff
@@ -0,0 +1,115 @@
+# Opennebula onesecgroup provider for Security Groups
+#
+# License: APLv2
+#
+# Authors:
+# Based upon initial work from Ken Barber
+# Modified by Martin Alfke
+#
+# Copyright
+# initial provider had no copyright
+# Deutsche Post E-POST Development GmbH - 2014, 2015
+#
+
+require 'rubygems'
+require 'nokogiri'
+
+Puppet::Type.type(:onesecgroup).provide(:cli) do
+  desc "onesecgroup provider"
+
+  has_command(:onesecgroup, "onesecgroup") do
+    environment :HOME => '/root', :ONE_AUTH => '/var/lib/one/.one/one_auth'
+  end
+
+  mk_resource_methods
+
+  # Create a security group with onesecgroup by passing in a temporary secgroup definition file.
+  def create
+    file = Tempfile.new("onesecgroup-#{resource[:name]}-create.xml")
+    builder = Nokogiri::XML::Builder.new do |xml|
+      xml.TEMPLATE do
+        xml.NAME resource[:name]
+        xml.DESCRIPTION resource[:description]
+        resource[:rules].each do |rule|
+          xml.RULE do
+            rule.each do |k, v|
+              xml.send(k.upcase, v)
+            end
+          end
+        end if resource[:rules]
+      end
+    end
+    tempfile = builder.to_xml
+    file.write(tempfile)
+    file.close
+    self.debug "Creating secgroup using #{tempfile}"
+    onesecgroup('create', file.path)
+    file.delete
+    @property_hash[:ensure] = :present
+  end
+
+  # Delete a secgroup using onesecgroup delete
+  def destroy
+    onesecgroup('delete', resource[:name])
+    @property_hash.clear
+  end
+
+  # Check if a secgroup exists by scanning the onesecgroup list
+  def exists?
+    @property_hash[:ensure] == :present
+  end
+
+  # Return the full hash of all existing secgroups
+  def self.instances
+    secgroups = Nokogiri::XML(onesecgroup('list', '-x')).root.xpath('/SECURITY_GROUP_POOL/SECURITY_GROUP')
```

Comment:
please check whether the following command works and delivers expected results:
puppet resource onesecgroup
This should create a Puppet DSL list of all available OpenNebula security groups.